### PR TITLE
Nisyad/add kvp adapter for vqa datasets

### PR DESCRIPTION
**Pull request by nisyad-ms**
This PR adds a a new adaptor which converts an existing VQA dataset to KVP format.

### CODE REVIEW

File: vision_datasets/visual_question_answering/vqa_as_kvp_dataset.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+        if self.QUESTION_KEY not in label:
+            raise KeyError(f"Question key '{self.QUESTION_KEY}' not found in label.")
+        if self.ANSWER_KEY not in label:
+            raise KeyError(f"Answer key '{self.ANSWER_KEY}' not found in label.")
+
+        return {
+            KeyValuePairLabelManifest.LABEL_KEY: {
+                self.ANSWER_KEY: {KeyValuePairLabelManifest.LABEL_VALUE_KEY: label[self.ANSWER_KEY]},
+                self.RATIONALE_KEY: {KeyValuePairLabelManifest.LABEL_VALUE_KEY: label[self.RATIONALE_KEY] if self.RATIONALE_KEY in label else ""}
```
Comment (cy-bai) [inline vision_datasets/visual_question_answering/vqa_as_kvp_dataset.py]: is the rationale availability consistent for all samples in a dataset? If so, we might add/delete the field in the schema level. If we have it in schema but some samples don't have rationale ground truth, the evaluator will consider any rational response as wrong

File: vision_datasets/visual_question_answering/vqa_as_kvp_dataset.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+            RATIONALE_KEY: {
+                "type": "string",
+                "description": "Rationale for the answer.",
+            },
+        }
+    }
+
+    def __init__(self, vqa_dataset: VisionDataset):
+        """
+        Initializes an instance of the ClassificationAsKeyValuePairDataset class.
```
Comment (cy-bai) [inline vision_datasets/visual_question_answering/vqa_as_kvp_dataset.py]: update comments

File: vision_datasets/visual_question_answering/vqa_as_kvp_dataset.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        Args:
+            classification_dataset (VisionDataset): The classification dataset to convert to key-value pair dataset.
```
Comment (cy-bai) [inline vision_datasets/visual_question_answering/vqa_as_kvp_dataset.py]: update comments

File: vision_datasets/visual_question_answering/vqa_as_kvp_dataset.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+class VQAAsKeyValuePairDataset(VisionDataset):
+    """Dataset class that access Visual Question Answering (VQA) datset as KeyValuePair dataset."""
+
+    ANSWER_KEY = "answer"
+    RATIONALE_KEY = "rationale"
+    QUESTION_KEY = "question"
+    SCHEMA_BASE = {
+        "name": "Visual Question Answering",
+        "description": "Answer questions on given images and provide rationale for the answer.",
```
Comment (cy-bai) [inline vision_datasets/visual_question_answering/vqa_as_kvp_dataset.py:22]: update description accordingly if rationale is not available / needed

File: vision_datasets/visual_question_answering/vqa_as_kvp_dataset.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+    def construct_schema(self) -> Dict[str, Any]:
+        return self.SCHEMA_BASE
+
+    def construct_kvp_label_data(self, label: Dict[str, str]) -> Dict[str, Dict[str, str]]:
+        """
+        Convert the classification dataset label_name to the desired format for KVP annotation as defined by the SCHEMA_BASE.
+        E.g. {"fields":
+                    {"answer": {"value": "yes"}},
+                    {"rationale": {"value": "I see a dog."}}
+             "text": {"question": "Is there a dog in the image?"}
```
Comment (cy-bai) [inline vision_datasets/visual_question_answering/vqa_as_kvp_dataset.py:84]: where is "question" filled? in this function?

File: vision_datasets/visual_question_answering/vqa_as_kvp_dataset.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+            raise KeyError(f"Question key '{self.QUESTION_KEY}' not found in label.")
+        if self.ANSWER_KEY not in label:
+            raise KeyError(f"Answer key '{self.ANSWER_KEY}' not found in label.")
+
+        return {
+            KeyValuePairLabelManifest.LABEL_KEY: {
+                self.ANSWER_KEY: {KeyValuePairLabelManifest.LABEL_VALUE_KEY: label[self.ANSWER_KEY]},
+                self.RATIONALE_KEY: {KeyValuePairLabelManifest.LABEL_VALUE_KEY: label[self.RATIONALE_KEY] if self.RATIONALE_KEY in label else ""}
+            },
+            KeyValuePairLabelManifest.TEXT_INPUT_KEY: {"question": label[self.QUESTION_KEY]},
```
Comment (cy-bai) [inline vision_datasets/visual_question_answering/vqa_as_kvp_dataset.py]: `"question"` should be `self.QUESTION_KEY`?

File: tests/test_ic_od_to_kvp_wrapper/test_vqa_as_kvp.py [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+            _, target, _ = kvp_dataset[0]
+            self.assertIsInstance(target, KeyValuePairLabelManifest)
+            self.assertEqual(target.label_data,
+                             {'fields': {'answer': {'value': 'answer 1'}, 'rationale': {'value': ''}}, 'text': {'question': 'question 1'}})
+
+            self.assertEqual(len(kvp_dataset), 3)
+            self.assertEqual(len(kvp_dataset.dataset_manifest.images), 2)
+
+            # Last image has 2 questions associated with it
+            self.assertEqual(kvp_dataset[-2][0][0].size, kvp_dataset[-1][0][0].size)
```
Comment (cy-bai) [inline tests/test_ic_od_to_kvp_wrapper/test_vqa_as_kvp.py]: It seems all images are the same with size (100, 100). Could we change the testcase and make the assert effective to check that the last two samples have the same image?

Comment (nisyad-ms) [reply tests/test_ic_od_to_kvp_wrapper/test_vqa_as_kvp.py]: Actually, I changed the logic to create image of different sizes for VQA to test exactly this. 
The second image is 200px and so on.

File: vision_datasets/visual_question_answering/vqa_as_kvp_dataset.py (hunk 3 above)
Comment (nisyad-ms) [reply vision_datasets/visual_question_answering/vqa_as_kvp_dataset.py]: ah good catch! updated

File: vision_datasets/visual_question_answering/vqa_as_kvp_dataset.py (hunk 4 above)
Comment (nisyad-ms) [reply vision_datasets/visual_question_answering/vqa_as_kvp_dataset.py:22]: updated

File: vision_datasets/visual_question_answering/vqa_as_kvp_dataset.py (hunk 2 above)
Comment (nisyad-ms) [reply vision_datasets/visual_question_answering/vqa_as_kvp_dataset.py]: updated

File: vision_datasets/visual_question_answering/vqa_as_kvp_dataset.py (hunk 1 above)
Comment (nisyad-ms) [reply vision_datasets/visual_question_answering/vqa_as_kvp_dataset.py]: "rationale" may not be present for some datasets.
I updated the `construct_kv_label_data` logic to only include "rationale" if it is present in the annotation sample. 
OR
Do you think not adding "rationale" at all makes more sense? I am not sure if we want to evaluate on "rationale" key?

File: vision_datasets/visual_question_answering/vqa_as_kvp_dataset.py [hunk 8]
```diff
@@ -0,0 +1,107 @@
+import logging
+from copy import deepcopy
```
Comment (juliagongms) [inline vision_datasets/visual_question_answering/vqa_as_kvp_dataset.py:1]: Nit: alphabetize imports - copy, logging, typing

**Review by juliagongms — COMMENTED** [review]
Looks good to me pending the remaining comments, thanks!

File: vision_datasets/visual_question_answering/vqa_as_kvp_dataset.py (hunk 1 above)
Comment (cy-bai) [reply vision_datasets/visual_question_answering/vqa_as_kvp_dataset.py]: yes, we can exclude it or add a boolean flag `rationale` and set to False by default.

**Comment by cy-bai** [comment]
add example code in README

File: vision_datasets/visual_question_answering/vqa_as_kvp_dataset.py (hunk 1 above)
Comment (nisyad-ms) [reply vision_datasets/visual_question_answering/vqa_as_kvp_dataset.py]: removed "rationale" alltogether for now. We can add back when needed.

**Review by cy-bai — APPROVED** [review]
LGTM, thanks

**Pull request merged**